pkg/dispatcher/dispatcher.go

```diff
@@ -100,16 +103,45 @@ func (p *Dispatcher) dispatch(event events.SchedulingEvent) error {
 	if !p.isRunning() {
 		return fmt.Errorf("dispatcher is not running")
 	}
-	for p.isRunning() {
-		select {
-		case p.eventChan <- event:
-			return nil
-		case <-time.After(time.Second * 3):
-			log.Logger.Error("event channel is full, keep waiting",
-				zap.Int("channelLength", len(p.eventChan)))
-		}
+	select {
+	case p.eventChan <- event:
+		log.Logger.Info("dispatched", zap.Int("capacity", cap(p.eventChan)),
+			zap.Int("length", len(p.eventChan)))
+		return nil
+	default:
+		log.Logger.Warn("event channel is full, transition to async-dispatch mode",
+			zap.Int("channelLength", len(p.eventChan)),
+			zap.String("event", fmt.Sprintf("%+v", event)))
+		p.asyncDispatch(event)
+		return nil
 	}
-	return fmt.Errorf("dispatcher is stopped")
+}
+
+// async-dispatch try to enqueue the event in every 3 seconds util timeout,
+// it's only called when event channel is full.
+func (p *Dispatcher) asyncDispatch(event events.SchedulingEvent) {
+	go func() {
+		beginTime := time.Now()
+		for p.isRunning() {
+			select {
+			case p.eventChan <- event:
+				return
+			case <-time.After(asyncDispatchCheckInterval):
+				elapseTime := time.Now().Sub(beginTime)
+				if elapseTime >= conf.GetSchedulerConf().DispatchTimeout {
+					log.Logger.Error("dispatch timeout",
+						zap.Float64("elapseSeconds", elapseTime.Seconds()),
+						zap.Int("channelLength", len(p.eventChan)),
+						zap.String("event", fmt.Sprintf("%+v", event)))
+					return
+				}
+				log.Logger.Warn("event channel is full, keep waiting...",
```

Comment:
the logging info should contain the time to wait here, something like:
`zap.String("retryingInSeconds", asyncDispatchCheckInterval)`

Reply:
Already done like this: `zap.Float64("elapseSeconds", elapseTime.Seconds())`.
Is it better to rename "elapseSeconds" as "retryingInSeconds" ?

Reply:
The elapse seconds that is logged shows the time since we have created the go routine, that is the correct thing to log. I do think that we need to log the `DispatchTimeout` but we should only log that on the start of the dispatcher as it cannot change while running (see previous remark on getting the config).

Also in this case: I don't think we need to know the channel length as it is still full otherwise we would not be here. Same for the event, type might be nice but we do not need the full event.

Reply:
Updates:
(1) Set beginTime before creating async-dispatch goroutines.
(2) Also remove 'event' and 'channelLength' fields from the log.